### PR TITLE
Control the number of times a change can be retried.

### LINKED ISSUE #93: Potential infinite loop with Channel.requeueChangeWithFullObject()
If requeueChangeWithFullObject() requeues a Change and an error is returned, it will keep trying to send until the cows come home.

### PR DESCRIPTION
**Pull request by roundhill**
Fixes #93.

### CODE REVIEW

**Comment by beaucollins** [comment]
Nice. Bonus points for a unit test.

@fredrocious if we receive a 405 after sending the full object, should we redownload the object?

**Comment by roundhill** [comment]
> Bonus points for a unit test.

I'm on it.

**Comment by chengfred** [comment]
@beaucollins I believe 405 after sending full object should not be possible. We should investigate if you think that's happened. Redownload whole object in that hypothetical though could be reasonable

**Comment by beaucollins** [comment]
We were seeing 405 responses to what _should_ have been changes with full JSON not just the diff. I can't confirm though that it was correctly sending the full JSON.

**Comment by roundhill** [comment]
This is the JSON it is attempting to send:

``` javascript
 {"id":"meta","v":{"last_seen":{"v":0,"o":"r"}},"sv":3453,"o":"M","d":{"last_seen":0,"latest_note_time":1397058587},"ccid":"2cd43fd69d9947a2938d1b8fccab893b"}
```

**Comment by beaucollins** [comment]
Looks correct to me. the `d` is the full object. The original change was just replacing the `last_seen` property to `0`.

File: Simperium/build.gradle [hunk 1]
```diff
@@ -19,6 +19,8 @@ repositories {
 dependencies {
     compile 'com.android:volley:4.4.2.1'
     compile 'com.codebutler:android-websockets:6c7c60d'
+
+    androidTestCompile 'com.jayway.android.robotium:robotium-solo:4.3.1'
```
Comment (beaucollins) [inline Simperium/build.gradle]: is this for running tests against a locally jdk instance?

**Comment by beaucollins** [comment]
@fredrocious looks like we're sending both the diff and the full object when adding full object data to a change message.

[Unit test is here](https://github.com/Simperium/simperium-android/blob/develop/Simperium/src/instrumentTest/java/com/simperium/ChannelTest.java#L638).

Example payload in the comment above:

``` json
 {"id":"meta","v":{"last_seen":{"v":0,"o":"r"}},"sv":3453,"o":"M","d":{"last_seen":0,"latest_note_time":1397058587},"ccid":"2cd43fd69d9947a2938d1b8fccab893b"}
```

File: Simperium/build.gradle (hunk 1 above)
Comment (roundhill) [inline Simperium/build.gradle]: Oh, maybe not needed. I was blindly following this guide, but we didn't need this dependency in the first place ;) http://tools.android.com/tech-docs/new-build-system/migrating_to_09

I'll remove it.

File: Simperium/src/androidTest/java/com/simperium/ChannelTest.java [hunk 2]
```diff
@@ -655,8 +649,22 @@ public void testResendChangeWithFullObjectAfterReceiving405()
...
+        note.setTitle("My plane is full of snakes");
+
+        Change change = new Change(Change.OPERATION_MODIFY, note);
+
+        // Spam a lot of attempts to requeue a change
+        for (int i=0; i < 10; i++) {
+            mChannel.requeueChangeWithFullObject(change);
+        }
 
+        assertEquals(Channel.RETRY_LIMIT, change.getRetryCount());
```
Comment (beaucollins) [inline Simperium/src/androidTest/java/com/simperium/ChannelTest.java:667]: `Change.getRetryCount()` is the number of times it attempted to actually send the retry out?

Comment (roundhill) [inline Simperium/src/androidTest/java/com/simperium/ChannelTest.java:667]: Yeah, it's just incrementing it every time it tries to requeue with the full upload again: https://github.com/Simperium/simperium-android/blob/issue/93-requeue-loop/Simperium/src/main/java/com/simperium/client/Channel.java#L352

**Comment by roundhill** [comment]
> @fredrocious looks like we're sending both the diff and the full object when adding full object data to a change message.

Aha, so it might processing the diff only and sending a 405 again?

**Comment by beaucollins** [comment]
I think sending both the diff and the full object is correct but just want to double check.

**Comment by beaucollins** [comment]
I'm going to try fixing #95 before merging this in.

**Comment by beaucollins** [comment]
I think this is good to merge then given the Simperium/poutine@cc593ca fix.

**Pull request merged**